srml/contract/src/tests.rs

```diff
@@ -506,3 +504,355 @@ fn dispatch_call() {
 		},
 	);
 }
+
+const CODE_SET_RENT: &str = r#"
+(module
+	(import "env" "ext_dispatch_call" (func $ext_dispatch_call (param i32 i32)))
+	(import "env" "ext_set_storage" (func $ext_set_storage (param i32 i32 i32 i32)))
+	(import "env" "ext_set_rent_allowance" (func $ext_set_rent_allowance (param i32 i32)))
+	(import "env" "ext_input_size" (func $ext_input_size (result i32)))
+	(import "env" "ext_input_copy" (func $ext_input_copy (param i32 i32 i32)))
+	(import "env" "memory" (memory 1 1))
+
+	(func $call_0
+		(call $ext_set_rent_allowance
+			(i32.const 0)
+			(i32.const 1)
+		)
+	)
+
+	(func $call_1
+		(call $ext_set_storage
+			(i32.const 1)
+			(i32.const 1)
+			(i32.const 0)
+			(i32.const 4)
+		)
+	)
+
+	(func $call_2
+		(call $ext_set_storage
+			(i32.const 1)
+			(i32.const 0)
+			(i32.const 0)
+			(i32.const 0)
+		)
+	)
+
+	(func $call_3
+		(call $ext_dispatch_call
+			(i32.const 8)
+			(i32.const 11)
+		)
+	)
+
+	(func $call_else)
+
+	(func $assert (param i32)
+		(block $ok
+			(br_if $ok
+				(get_local 0)
+			)
+			(unreachable)
+		)
+	)
+
+	(func (export "call")
+		(local $input_size i32)
+		(set_local $input_size
+			(call $ext_input_size)
+		)
+		(block $IF_3
+			(block $IF_2
+				(block $IF_1
+					(block $IF_0
+						(br_if $IF_0
+							(i32.eq
+								(get_local $input_size)
+								(i32.const 0)
+							)
+						)
+						(br_if $IF_1
+							(i32.eq
+								(get_local $input_size)
+								(i32.const 1)
+							)
+						)
+						(br_if $IF_2
+							(i32.eq
+								(get_local $input_size)
+								(i32.const 2)
+							)
+						)
+						(br_if $IF_3
+							(i32.eq
+								(get_local $input_size)
+								(i32.const 3)
+							)
+						)
+						(call $call_else)
+						return
+						(unreachable)
+					)
+					(call $call_0)
+					return
+					(unreachable)
+				)
+				(call $call_1)
+				return
+				(unreachable)
+			)
+			(call $call_2)
+			return
+			(unreachable)
```

Comment:
Here and above: I think `unreachable` is unnecessary here since `return` doesn't pass control past itself.